### PR TITLE
Adding Material Meta IDs to Segmentation Models

### PR DESCRIPTION
**Pull request by jordan-benjamin**
# Intro
The material of a block in Minecraft is described with a pair of numbers: a block ID and a meta ID. The block ID describes the high-level category that the material belongs to, such as stone or wood. The meta ID describes a specific material within a block-ID category - for example, a block granite and one of andesite will both share the same block ID (i.e. high-level category) of stone, but will have different meta IDs. A list of different materials and their corresponding IDs can be found here:

https://minecraft-ids.grahamedgecombe.com/

Currently, CraftAssist’s semantic segmentation model only uses the block ID of each material, thus ignoring some of the finer-grained differences between blocks. This PR adds the meta IDs to model training and inference. Experiments show that the new model converges significantly faster than the original model, and has both higher validation accuracy and loss. This PR also includes a few quality-of-life improvements for training, such as progress bars monitoring to monitor training, periodically outputting validation loss, and adding model accuracy code (thanks @snyxan!).

# Experiments
Attached to this PR are the notebooks I used comparing model training on the original dataset with the new data. To summarize the results, the new model reaches 98.44% validation accuracy after a single epoch, with small accuracy improvements in later epochs (although both train and validation loss continue to go down). Training the model on the original dataset, by comparison, results in 78% validation accuracy after 1 epoch, with the model reaching 97.7% over the course of 50 epochs.

[notebooks.zip](https://github.com/facebookresearch/craftassist/files/4705448/notebooks.zip)

# Some implementation details
- Inside this PR is a script that produces a meta-id-added segmentation dataset from the current segmentation dataset and the house dataset. The latter is needed because this is where each block’s meta ID is stored (they are already removed in the original segmentation dataset). Something to note here is that some houses in the segmentation dataset could not be found in the house dataset - however, this did not make the new dataset significantly smaller. 
- To turn (block_id, meta_id) pairs into a single material identifier, the entire segmentation dataset is iterated through to produce a material vocabulary, which maps pairs of IDs to arbitrary material IDs. This vocabulary is needed at inference time to convert (block, meta) ID pairs into the same material IDs that the model has been trained on.

# Next Steps:
- Perhaps instead of mapping each unique (block, meta) pair into totally separate material ids, and therefore entirely separate embedding vectors, a hybrid-embedding approach could be used, where each material’s embedding consists of a shared embedding coming from the block ID concatenated with a unique embedding for each meta ID. This would allow the model to learn a general representation of “stone” for example, and then make this more specific for each sub-material. This is probably overkill though - I think more impactful improvements would be in the architecture of the model.
- Now that the model can differentiate between similar materials, data augmentations could be added to, for example, swap all the blocks  of stone type A in a house with stone type B.
- The segmentation model itself could be experimented with. The current architecture doesn’t do any up or downsampling and has no residual connections - I’d like to try both (i.e. build a small UNET).

### CODE REVIEW

**Comment by snyxan** [comment]
Thanks for creating this PR! @jordan-benjamin 
Could you please point me to the script for creating/saving vocab file?

**Comment by jordan-benjamin** [comment]
Oops! Forgot to add it - see `python/craftassist/voxel_models/make_seg_ds.py`

**Comment by snyxan** [comment]
Thanks for uploading it! @jordan-benjamin 
LGTM.
